crates/services/block_aggregator_api/src/db/remote_cache.rs

```diff
@@ -0,0 +1,238 @@
+                .body(body)
+                .content_encoding("gzip")
+                .content_type("application/grpc-web");
+            let _ = req.send().await.map_err(Error::db_error)?;
+        }
+        match block_event {
+            BlockSourceEvent::NewBlock(new_height, _) => {
+                tracing::debug!("New block: {:?}", new_height);
+                self.highest_new_height = Some(new_height);
+                if self.synced {
+                    tracing::debug!("Updating latest block to {:?}", new_height);
+                    let mut tx = self.local_persisted.write_transaction();
+                    tx.storage_as_mut::<LatestBlock>()
+                        .insert(&(), &Mode::new_s3(new_height))
+                        .map_err(|e| Error::DB(anyhow!(e)))?;
+                    tx.commit().map_err(|e| Error::DB(anyhow!(e)))?;
+                } else if new_height == self.sync_from
+                    || self.height_is_next_height(new_height)?
+                {
+                    tracing::debug!("Updating latest block to {:?}", new_height);
+                    self.synced = true;
+                    self.highest_new_height = Some(new_height);
+                    self.orphaned_new_height = None;
+                    let mut tx = self.local_persisted.write_transaction();
+                    tx.storage_as_mut::<LatestBlock>()
+                        .insert(&(), &Mode::new_s3(new_height))
+                        .map_err(|e| Error::DB(anyhow!(e)))?;
+                    tx.commit().map_err(|e| Error::DB(anyhow!(e)))?;
+                } else if self.orphaned_new_height.is_none() {
+                    tracing::info!("Marking block as orphaned: {:?}", new_height);
+                    self.orphaned_new_height = Some(new_height);
+                }
+            }
+            BlockSourceEvent::OldBlock(height, _) => {
+                tracing::debug!("Old block: {:?}", height);
+                let mut tx = self.local_persisted.write_transaction();
+                let latest_height = if height.succ() == self.orphaned_new_height {
+                    tracing::debug!("Marking block as synced: {:?}", height);
+                    self.orphaned_new_height = None;
+                    self.synced = true;
+                    self.highest_new_height.unwrap_or(height)
+                } else {
+                    tracing::debug!("Updating latest block to {:?}", height);
+                    height
+                };
+                tx.storage_as_mut::<LatestBlock>()
+                    .insert(&(), &Mode::new_s3(latest_height))
+                    .map_err(|e| Error::DB(anyhow!(e)))?;
+                tx.commit().map_err(|e| Error::DB(anyhow!(e)))?;
+            }
+        }
+        Ok(())
+    }
+
+    async fn get_block_range(
+        &self,
+        first: BlockHeight,
+        last: BlockHeight,
+    ) -> crate::result::Result<Self::BlockRangeResponse> {
+        // TODO: Check if it exists
```

Comment:
Let's return an error in case we are not synced up to the requested height. But we don't need to check existance on S3, it will be too slow